### ISSUE TITLE
Are we moving towards including error messages in response body?

### ISSUE THREAD

**Issue by kristinapathak** [xmidt-org/httpaux#36]
I thought we were moving away from including any errors in the response body, but all encoders provided by `erraux` involve marshalling the error as `cause` into the response body:

default behavior:
https://github.com/xmidt-org/httpaux/blob/9f219068a591cf61a2c8deb156240876d76c9cef/erraux/encoder.go#L299-L305
`As` encoder:
https://github.com/xmidt-org/httpaux/blob/9f219068a591cf61a2c8deb156240876d76c9cef/erraux/encoder.go#L172-L173
https://github.com/xmidt-org/httpaux/blob/9f219068a591cf61a2c8deb156240876d76c9cef/erraux/encoder.go#L162
`Is` encoder:
https://github.com/xmidt-org/httpaux/blob/9f219068a591cf61a2c8deb156240876d76c9cef/erraux/encoder.go#L59
https://github.com/xmidt-org/httpaux/blob/9f219068a591cf61a2c8deb156240876d76c9cef/erraux/encoder.go#L112

Is this the plan for our API responses moving forward?

**Comment by johnabass** [comment]
I don't know about moving toward that.  However, it's a good, de facto standard to provide an error body.  The `erraux` package simply provides the same functionality as packages like `go-kit`, albeit in declarative fashion.

IIRC, @schmidtw didn't like the fact that some of our error bodies were plain text and some were JSON.  That, in turn, was due to us falling back to a plain text representation if the error wasn't recognized.

If we don't want error bodies, I can supply a setting in `erraux` to disable payloads.

**Comment by johnabass** [comment]
Just to drive home the point: https://tools.ietf.org/html/rfc7231#section-6.5

The 4xx (Client Error) class of status code indicates that the client
   seems to have erred.  Except when responding to a HEAD request, **the
   server SHOULD send a representation containing an explanation of the
   error situation**, and whether it is a temporary or permanent
   condition.  These status codes are applicable to any request method.
   User agents SHOULD display any included representation to the user.

(emphasis mine)

Note that that applies to the entire 4xx block, not just 400s.

**Comment by JC000** [comment]
It was, at one point, our collective deliberate decision to remove bodies from any HTTP errors.  This was largely in part to train our users to look at response codes rather than error bodies in any form when things went sideways.  Unless this decision has changed, we shouldn't change the behavior of the software as it is deployed in our environment.  If this behavior is easily changed via configuration, I'm all for adding the feature.

Ultimately though, this is a pointy hat boss decision.

**Comment by johnabass** [comment]
Perhaps the confusion is due to the fact that this repo is in `xmidt-org`?  My intention was for this to be a general purpose library, not something specific to the `xmidt` or `webpa` servers.  Would it make sense to move this library somewhere else?

I get (but was not part of the collective) that we do this on `xmidt`.  I would disagree that we want to ban all response bodies on errors on any project.  We should train our clients to read and adhere to the specification.

An example of an error response body being useful is `scrivener`.  It returns a 400 to indicate that the client did not send the correct name/value pairs for a template.  But which names were missing?  `Scrivener` indicates that in the response body.  If `scrivener` were not allowed to return a response body, then a client wouldn't be able to take remedial action which would likely include resorting to slack or filing a ticket.  The engineer implementing the codepath to hydrate a template won't always be on the same team as the content producer who authored the template.

Headers are inappropriate to communicate errors, as indicated by the RFC.  Among other reasons, headers are restricted to 7-bit ASCII and the error may include Unicode values.

**Comment by joe94** [comment]
Yeah, I'd vote for making it configurable as well. It'll help to at least start adopting the library and grow from there. 
FYI: Argus is reporting error messages through a header today but we can change that in later versions. We control the messages though and so far it's all ASCII.

**Comment by JC000** [comment]
What's a `scrivener`?  ;-)

@johnabass : I agree with you on all points.  I just want to make sure we don't change existing behavior.  Making the behavior configurable would satisfy both goals.

**Comment by schmidtw** [comment]
I'm in favor of making it configurable as well.  @JC000 summarized my position pretty well here: https://github.com/xmidt-org/httpaux/issues/36#issuecomment-811570460

While "SHOULD" is well meaning, when failure responses start containing payloads, consumers of the API may start decoding them & using them.  The payload becomes a poorly documented enforced API as sorts.  It becomes more fun when `content-type` and assorted `accept` headers also force the payload to be something you may not want.

**Comment by schmidtw** [comment]
To clarify my "dislike" ... I don't like getting "noise" defects because one layer of a system returns an `400` error as `applicatoin/json` with JSON vs. another section of code (maybe something we don't control) returns a simple text string for a different `400` error.  If we can normalize so the same response & format/etc can be more uniformly supported - great, I'm all for that.

**Comment by kristinapathak** [comment]
One other thing we might need to watch out for in our services is sending unfiltered errors to clients with information we don't want to expose, like in this argus issue:
https://github.com/xmidt-org/argus/issues/30
